### PR TITLE
Add support for latest release

### LINKED ISSUE #266: Support latest version tag
#### Is your feature request related to a problem?

There is no way to specify the latest version, such as `ubuntu-latest`. 

#### Describe the feature you'd like

Being able to retrieve the latest version of Elixir and Erlang OTP:
```
elixir:  latest
otp: latest
```

#### Describe alternatives you've considered

Currently using main and maint, but those branches are in constant development, and I want to be able to test on the latest release.

### PR DESCRIPTION
**Pull request by Schultzer**
# Description
This PR closes #266 and supersedes #267; where I got a bit lost in the codebase.

Closes #266

- [X] I have performed a self-review of my changes
- [X] I have read and understood the [contributing guidelines](/erlef/setup-beam/blob/main/CONTRIBUTING.md)

### CODE REVIEW

**Comment by paulo-ferraz-oliveira** [comment]
Haven't read the code yet, but I had an unanswered question in the other pull request.

Let's say `erlang/otp` releases 24.2.5, but 24.2.4 is the one set to "latest". Shall a rule in `setup-beam` override that? Or where will the semantics reside?

**Comment by Schultzer** [comment]
> Haven't read the code yet, but I had an unanswered question in the other pull request.
> 
> Let's say `erlang/otp` releases 24.2.5, but 24.2.4 is the one set to "latest". Shall a rule in `setup-beam` override that? Or where will the semantics reside?

Yeah, so if we want to provide a more true latest release then we need to do the reconciliation which would require a bigger refactoring, I tried walking down that path in the other PR but decided to bail out.

If there is a big interest in this we can certainly do it and we can leverage semver to do the heavy lifting by sorting the versions.

Regarding OTP and it’s branch versioning scheme, that can be treated as SemVer build meta and be sorted accordingly.

In the contribution guide it’s mentioned to keep changes to a minimum, so as a new contributor then I’m hesitant to rip apart most of this as it’s currently stand.

**Comment by Schultzer** [comment]
> > I have read and understood the [contributing guidelines](https://github.com/erlef/setup-beam/blob/main/CONTRIBUTING.md)
> 
> This states "Execute `npm run build-dist` and fix any issues arising from that". I can't understand what failed that it didn't generate any artefacts that you'd need to push, for which CI complained later.

I don't know how that didn't get pushed the first round, but I've rebased and regenerated the dist. Hopefully, it will pass this time around.

**Comment by paulo-ferraz-oliveira** [comment]
> In the contribution guide it’s mentioned to keep changes to a minimum, so as a new contributor then I’m hesitant to rip apart most of this as it’s currently stand.

I understand, and thank you for being patient and walking us through your thought process. The changes mostly seem sane to me (it's touching parts of the code that are somewhat sensitive), but I think I need other reviewers to pitch in.

In any case, tests are failing and it's probably due to `latest` being a moving target. You'll probably want to do static tests against a specific input list (and not reading from remote, as that'll surely fail as OTP keeps releasing patches and other versions on a given major branch).

**Comment by Schultzer** [comment]
> > In the contribution guide it’s mentioned to keep changes to a minimum, so as a new contributor then I’m hesitant to rip apart most of this as it’s currently stand.
> 
> I understand, and thank you for being patient and walking us through your thought process. The changes mostly seem sane to me (it's touching parts of the code that are somewhat sensitive), but I think I need other reviewers to pitch in.
> 
> In any case, tests are failing and it's probably due to `latest` being a moving target. You'll probably want to do static tests against a specific input list (and not reading from remote, as that'll surely fail as OTP keeps releasing patches and other versions on a given major branch).

Thanks, I've future proofed the tests.

File: test/setup-beam.test.js [hunk 1]
```diff
@@ -196,6 +196,11 @@ async function testOTPVersions() {
     expected = 'master'
     got = await setupBeam.getOTPVersion(spec, osVersion, hexMirrors)
     assert.deepStrictEqual(got, expected)
+
+    spec = 'latest'
+    osVersion = 'ubuntu-22.04'
+    got = await setupBeam.getOTPVersion(spec, osVersion, hexMirrors)
+    assert.match(got, /[.]/g)
```
Comment (paulo-ferraz-oliveira) [inline test/setup-beam.test.js:203]: Does this simply test you "got something"? Doesn't sound very robust. 😄 

In this "suite" you should find a list of versions that, wihout trying to download the artefacts check the output versus a pre-defined list: https://github.com/erlef/setup-beam/blob/main/test/setup-beam.test.js#L357

It'd be preferable to have the tests there, since the download mechanism is already well tested.

Comment (Schultzer) [reply test/setup-beam.test.js:203]: That test, is testing a function which is generic and to a degree that makes little sense to the project since some of the values are 1, 1.1 and etc. Those versions do not exist to my knowledge in any of the supported projects. As mentioned earlier it would require some heavy refactoring to that function, to get it to spit out the latest version, not impossible but the test values should be reconsidered. And ideally we only test valid versions of OTP, Elixir, Gleam, rebar and mix. Testing any potential versions makes little sense, since it would only introduce arbitrary rules which does not reflect reality.

Comment (paulo-ferraz-oliveira) [reply test/setup-beam.test.js:203]: I'm Ok if you create a set of versions that are SemVer and/or representative or the truth in the applications we support (Erlang, Elixir, Gleam, `rebar3`). What I don't see expressed, but maybe I missed it, is that the algo. is capable of choosing a pre-defined well-known "latest" from a given set of versions (this is what the tests require).

Comment (Schultzer) [reply test/setup-beam.test.js]: Cool. This was a bit of a learning experience for me, but I think it ticks all the boxes.

File: test/setup-beam.test.js [hunk 2]
```diff
@@ -542,7 +605,7 @@ async function testParseVersionFile() {
 erlang   ref:v${erlang}# comment, no space, and ref:v
 elixir ref:${elixir}  # comment, with space and ref:
  not-gleam 0.23 # not picked up
-gleam ${gleam}
```
Comment (paulo-ferraz-oliveira) [inline test/setup-beam.test.js:545]: This should be reversed. It's part of the test to make sure whitespace is not important in the parsing of the file.

File: test/setup-beam.test.js (hunk 1 above)
Comment (paulo-ferraz-oliveira) [reply test/setup-beam.test.js]: You went all the way 😄 What I meant was for you to either replicate or add to `https://github.com/erlef/setup-beam/blob/main/test/setup-beam.test.js#L357` 😄. You know the versions, so you know what your function should return; it'd be a minimal unit test (like the ones we already have). I mean, taking it from the source is also Ok, and it should be easy to keep those files updated, assuming you just copy-pasted them from their source (?)

**Comment by paulo-ferraz-oliveira** [comment]
Failing tests don't seem to be your responsibility; it's just that `rebar3` nightly stopped supporting OTP 24; I'll update that in a separate pull request.

**Comment by paulo-ferraz-oliveira** [comment]
@Schultzer, it'd be important to mention this in the README, with all the discussed constraints. Basically:

1. mention that `latest` is available
2. mention that it's calculated locally by the action, not directly related to what's set as `latest` in the remote repositories
3. if there's any constraints regarding x.y.z.a.b.c versions in Erlang/OTP also mention those as an exception

File: src/setup-beam.js [hunk 3]
```diff
@@ -379,6 +379,14 @@ function isStrictVersion() {
 }
 
 function getVersionFromSpec(spec0, versions0) {
+  let latest
+  Object.keys(versions0).forEach((version) => {
+    if (version.match(/main|master|nightly|latest/g) == null) {
```
Comment (paulo-ferraz-oliveira) [inline src/setup-beam.js:384]: Is this possibly missing `maint`?

File: test/setup-beam.test.js (hunk 1 above)
Comment (paulo-ferraz-oliveira) [reply test/setup-beam.test.js]: Fwiw: I really enjoy the latest straightforward and simple implementation.

**Comment by paulo-ferraz-oliveira** [comment]
When you squash your changes midway through a review it makes it harder on the reviewer to follow up 😕, since all the changes will now have to be reviewed again... No biggie, just letting you know how I feel it.

File: test/otp/ubuntu-22.04/builds.txt
Comment (paulo-ferraz-oliveira) [inline test/otp/ubuntu-22.04/builds.txt:1]: If you want to be complete here, there's already a `ubuntu-24.04` counterpart to this file...

File: test/setup-beam.test.js [hunk 4]
```diff
@@ -302,7 +343,6 @@ async function testElixirVersions() {
   got = await setupBeam.getElixirVersion(spec, otpVersion)
   assert.deepStrictEqual(got, expected)
   simulateInput('version-type', before)
-
```
Comment (paulo-ferraz-oliveira) [inline test/setup-beam.test.js:305]: This change should be reversed. It doesn't seem related to the pull request.

File: test/setup-beam.test.js [hunk 5]
```diff
@@ -526,6 +566,34 @@ async function testGetVersionFromSpec() {
...
+  spec = 'latest'
+  expected = '24.0.0'
+  got = setupBeam.getVersionFromSpec(spec, versions)
+  assert.deepStrictEqual(got, expected)
+
+  spec = 'latest'
+  got = setupBeam.getVersionFromSpec(spec, [
+    { '22.3.4.2.1.0.1': '22.3.4.2.1.0.1', '22.3.4.2.1.0.0': '22.3.4.2.1.0.0' },
+  ])
+  assert.deepStrictEqual(got, '22.3.4.2.1.0.1')
```
Comment (paulo-ferraz-oliveira) [inline test/setup-beam.test.js:579]: `'22.3.4.2.1.0.1'` should be a variable, for readability, as in the code above:

```
spec
expected
got
assert.
```

File: test/setup-beam.test.js [hunk 6]
```diff
@@ -526,6 +566,34 @@ async function testGetVersionFromSpec() {
...
+
+  got = setupBeam.getVersionFromSpec(spec, matrix.otp['ubuntu-18.04'])
+  assert.deepStrictEqual(got, '26.2.1')
```
Comment (paulo-ferraz-oliveira) [inline test/setup-beam.test.js:582]: As per the previous comment, the following structure should be adopted:
```
spec
expected
got
assert.
<newline>
```

(and you should repeat `spec` even if it's always `latest` so that the blocks make sense by themselves and we don't have to scroll to understand what we're testing)

File: test/setup-beam.test.js [hunk 7]
```diff
@@ -14,6 +14,47 @@ const core = require('@actions/core')
...
+      .forEach((x) => {
+        let v = x.name.match(/^otp_win64_(.*).exe$/)[1]
+        versions[v] = v
+      })
+    return versions
+  } else {
+    json.map((x) => (versions[x.tag_name] = x.tag_name))
+    return versions
+  }
+}
```
Comment (paulo-ferraz-oliveira) [inline test/setup-beam.test.js:56]: This could potentially be exported from the existing code base (`setup-beam.js`) but we can leave that as a future refactor.

File: test/setup-beam.test.js (hunk 1 above)
Comment (paulo-ferraz-oliveira) [reply test/setup-beam.test.js]: > 1, 1.1 and etc. Those versions do not exist to my knowledge in any of the supported projects

The versions themselves do not exist as releases, but `setup-beam` already infers the latest in a given "version branch". e.g. `24` could resolve to `24.3.4.17` so they make sense in the current implementation.

File: src/setup-beam.js (hunk 3 above)
Comment (Schultzer) [reply src/setup-beam.js]: `maint` is already being captured by `main`. I haven't made it to an array since I refactored it out when I made a more complete algorithm; the only limitation is for the version starting with `a` or `b,` which a couple of early versions for rebar3 do.

File: test/setup-beam.test.js (hunk 2 above)
Comment (Schultzer) [reply test/setup-beam.test.js:545]: Yeah, I'm not sure where or how to change this in Zed, and I assume it's a bug since this is happening in a string 

cc @zed-industries

File: test/otp/ubuntu-22.04/builds.txt
Comment (Schultzer) [reply test/otp/ubuntu-22.04/builds.txt:1]: Done.

**Comment by Schultzer** [comment]
> @Schultzer, it'd be important to mention this in the README, with all the discussed constraints. Basically:
> 
> 1. mention that `latest` is available
> 2. mention that it's calculated locally by the action, not directly related to what's set as `latest` in the remote repositories
> 3. if there's any constraints regarding x.y.z.a.b.c versions in Erlang/OTP also mention those as an exception

Done, no known constraints, should be all cake from here. 🍰

**Comment by Schultzer** [comment]
> When you squash your changes midway through a review it makes it harder on the reviewer to follow up 😕, since all the changes will now have to be reviewed again... No biggie, just letting you know how I feel it.

I apologize, the whitespace bit me, I couldn’t find the commit that I had it and I fat fingered git. Which ended up with me squeezing the changes. I’m a complete noob with regards to git.

**Comment by paulo-ferraz-oliveira** [comment]
@Schultzer, I don't think we have any more planned changes. I'm thus rebasing this onto the main branch to check if the error issues go away. If they do we can expect a squash+merge soon.

**Review by paulo-ferraz-oliveira — APPROVED** [review]

**Comment by paulo-ferraz-oliveira** [comment]
This is Ok ✅. I'm squash+merging. You can expect this change in the next release (we want to wait until #278 is done, so we can bump minor). Feel free to come back to this and ask for the release if it's taking more than, say, a week.
